### PR TITLE
Port synapse_port_db to async/await

### PR DESCRIPTION
**Pull request by babolivier**
Also raise an exception if there are pending background updates so we return with a non-0 code in the CI if something wrong happened when running the background updates.

Fixes https://github.com/matrix-org/synapse/issues/6719

### CODE REVIEW

**Comment by tadzik** [comment]
Running this locally still fails for me. As is:

```
Preparing sqlite3...
2020-01-16 21:56:47,381 - synapse_port_db - 601 - ERROR - 
Traceback (most recent call last):
  File "/home/synapse/synapse/env/bin/synapse_port_db", line 501, in run
    allow_outdated_version=True,
  File "/home/synapse/synapse/env/bin/synapse_port_db", line 471, in build_db_store
    db_conn, allow_outdated_version=allow_outdated_version
TypeError: check_database() got an unexpected keyword argument 'allow_outdated_version'
Traceback (most recent call last):
  File "/home/synapse/synapse/env/bin/synapse_port_db", line 501, in run
    allow_outdated_version=True,
  File "/home/synapse/synapse/env/bin/synapse_port_db", line 471, in build_db_store
    db_conn, allow_outdated_version=allow_outdated_version
TypeError: check_database() got an unexpected keyword argument 'allow_outdated_version'
```

Removing that parameter pushes it further, but it still dies:

```
Preparing sqlite3...
2020-01-16 21:57:45,737 - synapse.storage.prepare_database - 304 - INFO - Upgrading schema to v56
Preparing psycopg2...
2020-01-16 21:57:45,771 - synapse_port_db - 601 - ERROR - 
Traceback (most recent call last):
  File "/home/synapse/synapse/env/lib/python3.7/site-packages/twisted/internet/defer.py", line 1418, in _inlineCallbacks
    result = g.send(result)
StopIteration: []

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/synapse/synapse/env/lib/python3.7/site-packages/twisted/internet/defer.py", line 1418, in _inlineCallbacks
    result = g.send(result)
StopIteration: []

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/synapse/synapse/env/lib/python3.7/site-packages/twisted/internet/defer.py", line 1418, in _inlineCallbacks
    result = g.send(result)
StopIteration: True

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/synapse/synapse/env/bin/synapse_port_db", line 516, in run
    self.hs_config.get_single_database()
  File "/home/synapse/synapse/env/bin/synapse_port_db", line 471, in build_db_store
    db_conn
  File "/home/synapse/synapse/env/lib/python3.7/site-packages/synapse/storage/engines/postgres.py", line 36, in check_database
    txn.execute("SHOW SERVER_ENCODING")
AttributeError: 'psycopg2.extensions.connection' object has no attribute 'execute'
Traceback (most recent call last):
  File "/home/synapse/synapse/env/lib/python3.7/site-packages/twisted/internet/defer.py", line 1418, in _inlineCallbacks
    result = g.send(result)
StopIteration: []

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/synapse/synapse/env/lib/python3.7/site-packages/twisted/internet/defer.py", line 1418, in _inlineCallbacks
    result = g.send(result)
StopIteration: []

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/synapse/synapse/env/lib/python3.7/site-packages/twisted/internet/defer.py", line 1418, in _inlineCallbacks
    result = g.send(result)
StopIteration: True

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/synapse/synapse/env/bin/synapse_port_db", line 516, in run
    self.hs_config.get_single_database()
  File "/home/synapse/synapse/env/bin/synapse_port_db", line 471, in build_db_store
    db_conn
  File "/home/synapse/synapse/env/lib/python3.7/site-packages/synapse/storage/engines/postgres.py", line 36, in check_database
    txn.execute("SHOW SERVER_ENCODING")
AttributeError: 'psycopg2.extensions.connection' object has no attribute 'execute'
```

**Comment by babolivier** [comment]
@tadzik I looked into these logs and the corresponding code, it looks like it's caused by the changes in https://github.com/matrix-org/synapse/pull/6675 which haven't been released yet, hence why the script is working fine with a Synapse on the develop branch.

Given my local tests and the CI (which is now fixed, and passing), I'm pretty confident that the script now does the right thing.

So to try this patch, instead of blindly applying it onto your local version of `synapse_port_db`, there are two options:

* Setup a [development environment](https://github.com/matrix-org/synapse#synapse-development) for Synapse and checkout the `babolivier/port_db_bg_update_fix` branch
* Wait for the next release which will ship this fix (which should happen next week, or on the following one I think)

**Comment by tadzik** [comment]
Alright, having checked your branch (I'm at 75d60c2876e4f21f0ff18764d954631f2a18e38d), I'm still getting errors (though different ones now :)). They originate from synapse/storage/data_stores/main/schema/delta/57/local_current_membership.py, which I've modified as follows for debugging's sake:

```
--- a/synapse/storage/data_stores/main/schema/delta/57/local_current_membership.py
+++ b/synapse/storage/data_stores/main/schema/delta/57/local_current_membership.py
@@ -68,6 +68,8 @@ def run_upgrade(cur, database_engine, config, *args, **kwargs):
                 INNER JOIN room_memberships AS r USING (event_id)
                 WHERE type = 'm.room.member' and state_key like '%' || ?
         """
+    print(f"sql is '{sql}'")
+    print(f"config.server_name is '{config.server_name}'")
     cur.execute(sql, (config.server_name,))
 
     cur.execute(
```

The crash looks like so:

```
2020-01-17 13:25:39,547 - synapse.storage.prepare_database - 373 - INFO - Running script 57/local_current_membership.py
sql is '
            INSERT INTO local_current_membership (room_id, user_id, event_id, membership)
                SELECT c.room_id, state_key AS user_id, event_id, r.membership
                FROM current_state_events AS c
                INNER JOIN room_memberships AS r USING (event_id)
                WHERE type = 'm.room.member' and state_key like '%' || ?
        '
config.server_name is 'tadzik.net'
(...)
Traceback (most recent call last):
  File "/home/synapse/synapse/synapse.git/scripts/synapse_port_db", line 516, in run
    self.hs_config.get_single_database()
  File "/home/synapse/synapse/synapse.git/scripts/synapse_port_db", line 473, in build_db_store
    prepare_database(db_conn, engine, config=self.hs_config)
  File "/home/synapse/synapse/synapse.git/synapse/storage/prepare_database.py", line 85, in prepare_database
    data_stores=data_stores,
  File "/home/synapse/synapse/synapse.git/synapse/storage/prepare_database.py", line 376, in _upgrade_existing_database
    module.run_upgrade(cur, database_engine, config=config)
  File "/home/synapse/synapse/synapse.git/synapse/storage/data_stores/main/schema/delta/57/local_current_membership.py", line 73, in run_upgrade
    cur.execute(sql, (config.server_name,))
ValueError: unsupported format character ''' (0x27) at index 349
```

Some further debugging revealed that the offending character is the leading `'` in `state_key like '%'`. I don't see anything wrong with it myself, but perhaps you do :)

**Comment by babolivier** [comment]
Looks like that was https://github.com/matrix-org/synapse/issues/6727, fixed in https://github.com/matrix-org/synapse/pull/6728. I've merged the latest state of `develop` into my branch to pull the fix in, so feel free to `git pull` and try again.

**Comment by babolivier** [comment]
@tadzik just told me it worked and he's now live on PostgreSQL, so it looks like everything works correctly now :tada:

File: scripts/synapse_port_db [hunk 1]
```diff
@@ -505,21 +503,20 @@ class Porter(object):
...
-                yield self.sqlite_store.db.updates.has_completed_background_updates()
+                await self.sqlite_store.db.updates.has_completed_background_updates()
             )
             if not updates_complete:
-                sys.stderr.write(
+                raise Exception(
                     "Pending background updates exist in the SQLite3 database."
                     " Please start Synapse again and wait until every update has finished"
                     " before running this script.\n"
                 )
```
Comment (erikjohnston) [inline scripts/synapse_port_db]: Is there a way of quitting with a non-zero exit code that doesn't involve throwing an exception? I'd rather not have a big traceback in the logs, as opposed to nicely printing the rror directly to stderr

Comment (babolivier) [reply scripts/synapse_port_db]: There wasn't, but I fixed that in https://github.com/matrix-org/synapse/pull/6718/commits/8179108378f73f5ef72ce011da43685aefad31fb

File: scripts/synapse_port_db [hunk 2]
```diff
@@ -125,6 +129,7 @@ APPEND_ONLY_TABLES = [
 ]
 
 
+end_error = None
```
Comment (erikjohnston) [inline scripts/synapse_port_db]: Can you add a comment to what's going on here please

File: scripts/synapse_port_db [hunk 3]
```diff
@@ -489,13 +489,13 @@ class Porter(object):
...
-            postgres_ready = yield (
+            await self.postgres_store.db.updates.do_next_background_update(100)
+            postgres_ready = await (
                 self.postgres_store.db.updates.has_completed_background_updates()
             )
 
-    @defer.inlineCallbacks
-    def run(self):
+    async def run(self):
+        global end_error
```
Comment (erikjohnston) [inline scripts/synapse_port_db]: Can you also add a quick docstring for this that highlights that errors should be return via `end_error`?

**Pull request merged**